### PR TITLE
S3 storage

### PR DESCRIPTION
**Pull request by ChrisP19**

### CODE REVIEW

**Comment by ChrisP19** [comment]
Travis is failing because it doesn't have s3 access. Should I add the AWS keys as an encrypted environment variable as in https://docs.travis-ci.com/user/environment-variables/#defining-encrypted-variables-in-travisyml ?

File: app/src/js/server/s3_storage.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  /**
+   * Init bucket
+   */
+  public async makeBucket (): Promise<void> {
+    // create new bucket if there isn't one already (wait until it exists)
+    const hasBucket = await this.hasBucket()
+    if (!hasBucket) {
+      try {
```
Comment (fyu) [inline app/src/js/server/s3_storage.ts]: try to remove the try block?

File: app/src/js/server/s3_storage.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  public async save (key: string, json: string): Promise<void> {
+    const params = {
+      Body: json,
+      Bucket: this.bucketName,
+      Key: this.fullFile(key)
+    }
+    const savePromise = this.s3.putObject(params).promise()
+    return savePromise.then(() => {
+      return
+    })
```
Comment (fyu) [inline app/src/js/server/s3_storage.ts]: `return await this.s3.putObject(params).promise()`?

File: app/src/js/server/s3_storage.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+   * Loads fields stored at a key
+   * @param {string} key: relative path of file
+   */
+  public async load (key: string): Promise<string> {
+    const params = {
+      Bucket: this.bucketName,
+      Key: this.fullFile(key)
+    }
+    try {
+      const data = (await this.s3.getObject(params).promise()).Body
```
Comment (fyu) [inline app/src/js/server/s3_storage.ts]: Check the return promised value first

File: app/src/js/server/util.ts [hunk 4]
```diff
@@ -38,37 +39,40 @@ export function initEnv () {
...
         '%s storage not implemented yet, using file storage', database))
-      break
+      return [err, new FileStorage(dir)]
     }
     case DatabaseType.LOCAL: {
-      break
+      return [null, new FileStorage(dir)]
     }
     default: {
-      err = Error(sprintf(
```
Comment (fyu) [inline app/src/js/server/util.ts]: Gentle error handling

**Pull request merged**